src/shared/listener-collection.js

```diff
@@ -11,24 +11,41 @@
  */
 export class ListenerCollection {
   constructor() {
-    /** @type {Listener[]} */
+    /** @type {Array<Listener|null>} */
     this._listeners = [];
   }
 
   /**
+   * Adds a listener and returns a listenerId
    * @param {Listener['eventTarget']} eventTarget
    * @param {Listener['eventType']} eventType
    * @param {Listener['listener']} listener
    * @param {AddEventListenerOptions} [options]
    */
   add(eventTarget, eventType, listener, options) {
     eventTarget.addEventListener(eventType, listener, options);
-    this._listeners.push({ eventTarget, eventType, listener });
+    return this._listeners.push({ eventTarget, eventType, listener }) - 1;
   }
 
-  removeAll() {
-    this._listeners.forEach(({ eventTarget, eventType, listener }) => {
+  /**
+   * Removes a listener using a listenerId
+   * @param {number} listenerId
+   */
+  remove(listenerId) {
+    const event = this._listeners[listenerId];
+    if (event) {
+      const { eventTarget, eventType, listener } = event;
       eventTarget.removeEventListener(eventType, listener);
+      this._listeners[listenerId] = null;
```

Comment:
I see steps have been taken to avoid ID reuse here, which is good. It does mean that you have a growing array of empty slots to loop over in future though. Also `removeAll` will result in ID reuse.